tests/model_inheritance/test_abstract_inheritance.py

```diff
@@ -1,63 +1,106 @@
+        """
+        self._build_class('AbstractBase', None, 30, abstract=True)
+        self._build_class('AbstractDescendantNO', ['AbstractBase'], abstract=True)
+        self._build_class('AbstractDescendantO', ['AbstractBase'], 40, abstract=True)
+        self._build_class('DerivedFromBaseNO', ['AbstractBase'])
+        self._build_class('DerivedFromBaseO', ['AbstractBase'], 50)
+        self._build_class('DerivedFromDescendantNONO', ['AbstractDescendantNO'])
+        self._build_class('DerivedFromDescendantNOO', ['AbstractDescendantNO'], 50)
+        self._build_class('DerivedFromDescendantONO', ['AbstractDescendantO'])
+        self._build_class('DerivedFromDescendantOO', ['AbstractDescendantO'], 50)
+
+        self.assertCorrectInheritance('DerivedFromBaseNO', 30)
+        self.assertCorrectInheritance('DerivedFromBaseO', 50)
+        self.assertCorrectInheritance('DerivedFromDescendantNONO', 30)
+        self.assertCorrectInheritance('DerivedFromDescendantNOO', 50)
+        self.assertCorrectInheritance('DerivedFromDescendantONO', 40)
+        self.assertCorrectInheritance('DerivedFromDescendantOO', 50)
+
+    def test_multi_parent(self):
+        """ Test mro preserved when inheriting from multiple parents at once.
+            NO = non-overriding, O = overriding.
+        """
+        self._build_class('AbstractBase1', None, 30, abstract=True)
+        self._build_class('AbstractBase2', None, 40, abstract=True)
+        self._build_class('AbstractBase3', None, None, abstract=True)
+        self._build_class('DerivedFromBases12NO', ['AbstractBase1', 'AbstractBase2'])
+        self._build_class('DerivedFromBases13NO', ['AbstractBase1', 'AbstractBase3'])
+        self._build_class('DerivedFromBases32NO', ['AbstractBase3', 'AbstractBase2'])
+        self._build_class('DerivedFromBases12O', ['AbstractBase1', 'AbstractBase2'], 50)
+        self._build_class('DerivedFromBases13O', ['AbstractBase1', 'AbstractBase3'], 50)
+        self._build_class('DerivedFromBases32O', ['AbstractBase3', 'AbstractBase2'], 50)
+
+        self.assertCorrectInheritance('DerivedFromBases12NO', 30)
+        self.assertCorrectInheritance('DerivedFromBases13NO', 30)
+        self.assertCorrectInheritance('DerivedFromBases32NO', 40)
+        self.assertCorrectInheritance('DerivedFromBases12O', 50)
+        self.assertCorrectInheritance('DerivedFromBases13O', 50)
+        self.assertCorrectInheritance('DerivedFromBases32O', 50)
+
+    @expectedFailure
+    def test_multi_parent_diamond(self):
+        """ Test mro preserved when inheriting from multiple parents in diamond formation.
+            NO = non-overriding, O = overriding.
+
+            NOTE: We need a rewrite of `ModelBase.__new__()` to support this properly.
+            It's marked as expected failure for now, and because we can't control
+            the inheritance the way we'd like to, we should block users from building
+            diamond cases at all, rather than testing for the converse.
```

Comment:
> we should block users from building diamond cases at all, rather than testing for the converse.

But we don't block, right? It's not normally worth leaving comments about things Django "should" do.

Also I think this test can be split into a few different diamond cases.

Reply:
> > we should block users from building diamond cases at all, rather than testing for the converse.
> 
> But we don't block, right? It's not normally worth leaving comments about things Django "should" do.

Yeah fair. I'll update this comment once we decide whether we go ahead and address the `should` bit prior to merging this PR.

> Also I think this test can be split into a few different diamond cases.

There are 4 cases (A=Abstract, *=defines field). I think all are worth checking, based on :
```
  A*     A*     A*     A*
 / \    / \    / \    / \ 
A*  A  A*  A  A   A* A   A*
 \ /    \ /    \ /    \ /  
  D*     D      D*     D
```

So four separate test cases for diamond inheritance? Feels very verbose. Possibly no need to test in cases where `Derived` overrides with its own value since the earlier multi inheritance test already covers this, but given the complexity of the underlying metaclass logic, might be worth keeping still?

Reply:
I think it's fine. Separate test cases are normally a good thing since they can fail individually.